components/camel-kubernetes/src/main/java/org/apache/camel/component/kubernetes/events/KubernetesEventsConsumer.java

```diff
@@ -0,0 +1,140 @@
+    private EventsConsumerTask eventWatcher;
+
+    public KubernetesEventsConsumer(AbstractKubernetesEndpoint endpoint, Processor processor) {
+        super(endpoint, processor);
+        this.processor = processor;
+    }
+
+    @Override
+    public AbstractKubernetesEndpoint getEndpoint() {
+        return (AbstractKubernetesEndpoint) super.getEndpoint();
+    }
+
+    @Override
+    protected void doStart() throws Exception {
+        super.doStart();
+        executor = getEndpoint().createExecutor();
+
+        eventWatcher = new EventsConsumerTask();
+        executor.submit(eventWatcher);
+    }
+
+    @Override
+    protected void doStop() throws Exception {
+        super.doStop();
+
+        LOG.debug("Stopping Kubernetes Event Consumer");
+        if (executor != null) {
+            KubernetesHelper.close(eventWatcher, eventWatcher::getWatch);
+
+            if (getEndpoint() != null && getEndpoint().getCamelContext() != null) {
+                getEndpoint().getCamelContext().getExecutorServiceManager().shutdownNow(executor);
+            } else {
+                executor.shutdownNow();
+            }
+        }
+        executor = null;
+    }
+
+    class EventsConsumerTask implements Runnable {
+
+        private Watch watch;
+
+        @Override
+        public void run() {
+
+            FilterWatchListDeletable<Event, EventList, Resource<Event>> w = null;
+            if (ObjectHelper.isNotEmpty(getEndpoint().getKubernetesConfiguration().getLabelKey())
+                    && ObjectHelper.isNotEmpty(getEndpoint().getKubernetesConfiguration().getLabelValue())) {
+                w = getEndpoint().getKubernetesClient().events().v1().events().withLabel(
+                        getEndpoint().getKubernetesConfiguration().getLabelKey(),
+                        getEndpoint().getKubernetesConfiguration().getLabelValue());
+            }
+            if (ObjectHelper.isNotEmpty(getEndpoint().getKubernetesConfiguration().getResourceName())) {
+                Resource<Event> eventResource = getEndpoint()
+                        .getKubernetesClient().events().v1().events()
+                        .withName(getEndpoint().getKubernetesConfiguration().getResourceName());
+                w = (FilterWatchListDeletable<Event, EventList, Resource<Event>>) eventResource;
+            }
+            if (w == null) {
+                throw new RuntimeCamelException("Consumer label key or consumer resource name need to be set.");
+            }
+            watch = w.watch(new Watcher<>() {
+
+                @Override
+                public void eventReceived(Action action, Event resource) {
+                    Exchange exchange = createExchange(false);
+                    exchange.getIn().setBody(resource);
+                    exchange.getIn().setHeader(KubernetesConstants.KUBERNETES_EVENT_ACTION, action);
+                    exchange.getIn().setHeader(KubernetesConstants.KUBERNETES_EVENT_TIMESTAMP, System.currentTimeMillis());
```

Comment:
Just out of curiosity: wouldn't it be better/possible to use a more elaborate data type for the timestamp? 

My concern here is that this data may be confusing / hard to work with in scenarios where events may happen in different timezone.

Would it make sense or is it irrelevant in this context?

Reply:
This timestamp is actually the same as other resource types, for the sake of consistency, I think that we should keep the same type. BTW, please note that you have a more specific type in the event itself stored in the message body.

Reply:
Thanks for clarifying!